cmd/object-api-input-checks.go

```diff
@@ -0,0 +1,151 @@
+		return traceError(ObjectNameInvalid{Bucket: bucket, Object: object})
+	}
+	return nil
+}
+
+// Checks for all ListObjects arguments validity.
+func checkListObjsArgs(bucket, prefix, marker, delimiter string, obj ObjectLayer) error {
+	// Verify bucket exists.
+	if err := checkBucketExist(bucket, obj); err != nil {
+		return traceError(err)
+	}
+	if !IsValidObjectPrefix(prefix) {
+		return traceError(ObjectNameInvalid{
+			Bucket: bucket,
+			Object: prefix,
+		})
+	}
+	// Verify if delimiter is anything other than '/', which we do not support.
+	if delimiter != "" && delimiter != slashSeparator {
+		return traceError(UnsupportedDelimiter{
+			Delimiter: delimiter,
+		})
+	}
+	// Verify if marker has prefix.
+	if marker != "" && !strings.HasPrefix(marker, prefix) {
+		return traceError(InvalidMarkerPrefixCombination{
+			Marker: marker,
+			Prefix: prefix,
+		})
+	}
+	return nil
+}
+
+// Checks for all ListMultipartUploads arguments validity.
+func checkListMultipartArgs(bucket, prefix, keyMarker, uploadIDMarker, delimiter string, obj ObjectLayer) error {
+	if err := checkListObjsArgs(bucket, prefix, keyMarker, delimiter, obj); err != nil {
+		return err
+	}
+	if uploadIDMarker != "" {
+		if strings.HasSuffix(keyMarker, slashSeparator) {
+			return traceError(InvalidUploadIDKeyCombination{
+				UploadIDMarker: uploadIDMarker,
+				KeyMarker:      keyMarker,
+			})
+		}
+		id, err := uuid.Parse(uploadIDMarker)
+		if err != nil {
+			return traceError(err)
+		}
+		if id.IsZero() {
+			return traceError(MalformedUploadID{
+				UploadID: uploadIDMarker,
+			})
+		}
+	}
+	return nil
+}
+
+// Checks for NewMultipartUpload arguments validity, also validates if bucket exists.
+func checkNewMultipartArgs(bucket, object string, obj ObjectLayer) error {
+	return checkPutObjectArgs(bucket, object, obj)
+}
+
+// Checks for PutObjectPart arguments validity, also validates if bucket exists.
+func checkPutObjectPartArgs(bucket, object string, obj ObjectLayer) error {
+	return checkPutObjectArgs(bucket, object, obj)
+}
+
+// Checks for ListParts arguments validity, also validates if bucket exists.
+func checkListPartsArgs(bucket, object string, obj ObjectLayer) error {
+	return checkPutObjectArgs(bucket, object, obj)
+}
+
+// Checks for CompleteMultipartUpload arguments validity, also validates if bucket exists.
+func checkCompleteMultipartArgs(bucket, object string, obj ObjectLayer) error {
+	return checkPutObjectArgs(bucket, object, obj)
+}
+
+// Checks for AbortMultipartUpload arguments validity, also validates if bucket exists.
+func checkAbortMultipartArgs(bucket, object string, obj ObjectLayer) error {
+	return checkPutObjectArgs(bucket, object, obj)
+}
+
+// Checks for PutObject arguments validity, also validates if bucket exists.
+func checkPutObjectArgs(bucket, object string, obj ObjectLayer) error {
+	// Verify bucket exists.
+	if err := checkBucketExist(bucket, obj); err != nil {
+		return traceError(err)
+	}
+	if !IsValidObjectName(object) {
```

Comment:
`IsValidObjectName` is a simpler check (involves only string processing) and it should be done before `checkBucketExist` as that is a more expensive object layer check.